### PR TITLE
Enable querying time offset metrics with `agg_time_dimension`

### LINKED ISSUE #1006: [SL-1611] [Feature] Enable querying time offset metrics with their agg_time_dimension
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing metricflow functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Currently, time offset metrics can only be queried with `metric_time`. Since metric_time represents the `agg_time_dimensions` associated with the metric's measures, you should be able to replace `metric_time` with the `agg_time_dimension` and get the same behavior.
Note: if a metric has multiple input metrics with different `agg_time_dimensions`, `metric_time` will still be required.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

_No response_

<sub>[SL-1611](https://linear.app/dbt-labs/issue/SL-1611/[feature]-enable-querying-time-offset-metrics-with-their-agg-time)</sub>

### PR DESCRIPTION
**Pull request by courtneyholcomb**
Resolves #1006

### Description
Enable querying time offset metrics with their `agg_time_dimension` in place of `metric_time`.
Should only work for when all input metrics have the same `agg_time_dimension`.

### CODE REVIEW

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 1]
```diff
@@ -505,12 +505,20 @@ def _build_derived_metric_output_node(
...
+                    metric_spec.reference
+                )
+                queried_agg_time_dimension_specs = list(
+                    set(queried_linkable_specs.time_dimension_specs).intersection(set(valid_agg_time_dimensions))
+                )
             assert (
-                queried_linkable_specs.contains_metric_time
-            ), "Joining to time spine requires querying with metric_time."
+                queried_agg_time_dimension_specs
+            ), "Joining to time spine requires querying with metric_timeor the appropriate agg_time_dimension."
```
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:518]: nit `metric_time or`

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 2]
```diff
@@ -787,9 +795,7 @@ def _build_measure_spec_properties(self, measure_specs: Sequence[MeasureSpec]) -
                 f"semantic models: {semantic_models}. This suggests the measure_specs were not correctly filtered."
             )
 
-        agg_time_dimension = agg_time_dimension = self._semantic_model_lookup.get_agg_time_dimension_for_measure(
```
Comment (tlento) [inline metricflow/dataflow/builder/dataflow_plan_builder.py:790]: LOL. I did not notice that.

File: metricflow/plan_conversion/dataflow_to_sql.py [hunk 3]
```diff
@@ -1244,56 +1245,77 @@ def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode) -> SqlDataSet
...
-        ), "Can't join to time spine without metric time. Validations should have prevented this."
-        metric_time_dimension_column_name = self.column_association_resolver.resolve_spec(
-            metric_time_dimension_instance.spec
+            len(node.requested_agg_time_dimension_specs) > 0
+        ), "Must have at least one value in requested_agg_time_dimension_specs for JoinToTimeSpineNode."
+
+        # Determine if the time spine join should use metric_time or the agg_time_dimension (metric_time takes priority).
+        agg_time_dimension_for_join = node.requested_agg_time_dimension_specs[0]
+        for spec in node.requested_agg_time_dimension_specs[1:]:
+            if spec.element_name == METRIC_TIME_ELEMENT_NAME:
```
Comment (tlento) [inline metricflow/plan_conversion/dataflow_to_sql.py:1255]: This is no longer guaranteed to be the lowest granularity, right? Is that intentional? It seems like it doesn't matter, but it's not clear at this point if that's really the case or not.

File: metricflow/plan_conversion/dataflow_to_sql.py [hunk 4]
```diff
@@ -1244,56 +1245,77 @@ def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode) -> SqlDataSet
...
-            metric_time_dimension_instance
-        ), "Can't join to time spine without metric time. Validations should have prevented this."
-        metric_time_dimension_column_name = self.column_association_resolver.resolve_spec(
-            metric_time_dimension_instance.spec
+            len(node.requested_agg_time_dimension_specs) > 0
+        ), "Must have at least one value in requested_agg_time_dimension_specs for JoinToTimeSpineNode."
+
+        # Determine if the time spine join should use metric_time or the agg_time_dimension (metric_time takes priority).
+        agg_time_dimension_for_join = node.requested_agg_time_dimension_specs[0]
+        for spec in node.requested_agg_time_dimension_specs[1:]:
```
Comment (tlento) [inline metricflow/plan_conversion/dataflow_to_sql.py:1254]: This is a little magical - how do we know the first one is always not-metric-time and the others are always metric time? What if the user just requested metric_time, wouldn't the first one be potentially lowest-granularity metric_time?

I think it'll be cleaner to add a property to the node that gives us back the spec we need here, then we can be intentional with the naming and establish what people should care about. Like if all we need is the name/link pattern of an agg time dimension spec we can even return just those things.

File: metricflow/plan_conversion/dataflow_to_sql.py [hunk 5]
```diff
@@ -1244,56 +1245,77 @@ def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode) -> SqlDataSet
...
-                    instance.spec.time_granularity < metric_time_dimension_instance.spec.time_granularity
-                ):
-                    metric_time_dimension_instance = instance
         assert (
-            metric_time_dimension_instance
-        ), "Can't join to time spine without metric time. Validations should have prevented this."
-        metric_time_dimension_column_name = self.column_association_resolver.resolve_spec(
-            metric_time_dimension_instance.spec
+            len(node.requested_agg_time_dimension_specs) > 0
+        ), "Must have at least one value in requested_agg_time_dimension_specs for JoinToTimeSpineNode."
```
Comment (tlento) [inline metricflow/plan_conversion/dataflow_to_sql.py:1250]: Can we put this assertion in the JoinToTimeSpineNode? The return type is Sequence so we shouldn't be able to append to it, but do we ever update that property in other ways?

File: metricflow/query/validation_rules/metric_time_requirements.py [hunk 6]
```diff
@@ -65,6 +46,18 @@ def _group_by_items_include_metric_time(self, query_resolver_input: ResolverInpu
 
         return False
 
+    def _group_by_items_include_agg_time_dimension(
+        self, query_resolver_input: ResolverInputForQuery, metric_reference: MetricReference
```
Comment (tlento) [inline metricflow/query/validation_rules/metric_time_requirements.py:50]: This change was made in the other PR too, right?

**Review by tlento — COMMENTED** [review]
The main thing I'm seeing here is that this logic is kind of the same everywhere.

In this case, the two tests I think we'd like to have are:

1. Rendering the case where the metric uses multiple measures with the same agg_time_dimension (if that isn't already here, kind of hard to tell)
2. Throwing the error in the case where the metric uses mulitple agg_time_dimensions

I'm assuming case 2 should throw here as well.

If we're able to factor out the repeated logic we may be able to test the conditions that component handles directly and then keep our overall query rendering test suite a little more streamlined, but it's not clear how viable that is.

File: metricflow/dataflow/builder/dataflow_plan_builder.py [hunk 7]
```diff
@@ -505,12 +505,20 @@ def _build_derived_metric_output_node(
...
         # For nested ratio / derived metrics with time offset, apply offset & where constraint after metric computation.
         if metric_spec.has_time_offset:
+            queried_agg_time_dimension_specs = list(queried_linkable_specs.metric_time_specs)
+            if not queried_agg_time_dimension_specs:
+                valid_agg_time_dimensions = self._metric_lookup.get_valid_agg_time_dimensions_for_metric(
+                    metric_spec.reference
+                )
+                queried_agg_time_dimension_specs = list(
+                    set(queried_linkable_specs.time_dimension_specs).intersection(set(valid_agg_time_dimensions))
+                )
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/dataflow/builder/dataflow_plan_builder.py:515]: Was planning to resolve this in merge conflict resolution once one of the PRs merges, but I like the idea of adding a property to make this cleaner 🙏🏼 (My initial PR had the logic consolidated, but I ended up duplicating when I split into smaller PRs.)

File: metricflow/plan_conversion/dataflow_to_sql.py (hunk 4 above)
Comment (courtneyholcomb) [reply metricflow/plan_conversion/dataflow_to_sql.py:1254]: We don't know if the first one is always not-metric-time, I just overlooked checking the first one 🤦 

Re: granularity - the purpose of this check is just to see if the user passed `metric_time` or `agg_time_dimension` - it doesn't care about the granularity, so it stops as soon as it finds ANY `metric_time`. The granularity check happens a few lines down when going through all time dimension instances in the parent data set.

But I agree with your point about more intentional naming - will clean that up!

File: metricflow/query/validation_rules/metric_time_requirements.py (hunk 6 above)
Comment (courtneyholcomb) [reply metricflow/query/validation_rules/metric_time_requirements.py:50]: Right. Planning to resolve that in merge conflict resolution.

Comment (courtneyholcomb) [reply metricflow/query/validation_rules/metric_time_requirements.py:50]: Next time will probably base the branches off of each other instead... Didn't realize they still had overlap when I made that choice.

File: metricflow/dataflow/builder/dataflow_plan_builder.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/dataflow/builder/dataflow_plan_builder.py]: Updated!

File: metricflow/plan_conversion/dataflow_to_sql.py (hunk 3 above)
Comment (courtneyholcomb) [reply metricflow/plan_conversion/dataflow_to_sql.py:1255]: The granularity check happens a few lines down!

File: metricflow/plan_conversion/dataflow_to_sql.py (hunk 4 above)
Comment (courtneyholcomb) [reply metricflow/plan_conversion/dataflow_to_sql.py]: Alright, I added a new property to the node called `query_contains_metric_time` and removed this logic!

**Review by tlento — COMMENTED** [review]
Submitting commentary for now, be back later. :)

File: metricflow/dataflow/dataflow_plan.py [hunk 8]
```diff
@@ -673,6 +673,7 @@ def __init__(
         self,
         parent_node: BaseOutput,
         requested_agg_time_dimension_specs: List[TimeDimensionSpec],
+        query_includes_metric_time: bool,
```
[the remark this answers is not in the document]
Comment (courtneyholcomb) [reply metricflow/dataflow/dataflow_plan.py]: Updated!

**Comment by courtneyholcomb** [comment]
> In this case, the two tests I think we'd like to have are:
> 
> Rendering the case where the metric uses multiple measures with the same agg_time_dimension (if that isn't already here, kind of hard to tell)
> Throwing the error in the case where the metric uses mulitple agg_time_dimensions

Added these tests! See the two latest commits.

File: metricflow/test/query/test_query_parser.py [hunk 9]
```diff
@@ -584,3 +608,16 @@ def test_duplicate_metric_query(bookings_query_parser: MetricFlowQueryParser) ->
...
         bookings_query_parser.parse_and_validate_query()
+
+
+def test_offset_metric_with_diff_agg_time_dims_error() -> None:  # noqa: D
+    metrics_yaml_file = YamlConfigFile(filepath="inline_for_test_1", contents=METRICS_YAML)
+    revenue_yaml_file = YamlConfigFile(filepath="inline_for_test_1", contents=REVENUE_YAML)
+    query_parser = query_parser_from_yaml(
+        [EXAMPLE_PROJECT_CONFIGURATION_YAML_CONFIG_FILE, revenue_yaml_file, metrics_yaml_file]
+    )
+    with pytest.raises(InvalidQueryException, match="does not match any of the available group-by-items"):
```
Comment (tlento) [inline metricflow/test/query/test_query_parser.py:619]: Out of curiosity, does this raise a more specific metric-time related error somewhere? Might be a nice additional improvement.

**Review by tlento — APPROVED** [review]
Huzzah!

**Pull request merged**